system/jhtree/jhtree.cpp

```diff
@@ -1472,29 +1472,92 @@ void CDiskKeyIndex::ensureReady()
+
+    const bool optimisticTailRead = true;
+    bool foundTrailingHeader = false;
     KeyHdr hdr;
-    offset_t sizeRead = io->read(0, sizeof(hdr), &hdr);
-    if (sizeRead != sizeof(hdr))
-        throw MakeStringException(0, "Failed to read key '%s' header: file too small, could not read %u bytes - read %u", name.str(), (unsigned) sizeof(hdr), (unsigned)sizeRead);
 
-#ifdef _DEBUG
-    //In debug mode always use the trailing header if it is available to ensure that code path is tested
-    if (hdr.ktype & USE_TRAILING_HEADER)
-#else
-    if (hdr.ktype & TRAILING_HEADER_ONLY)
-#endif
+    // Read the end of the file, and walk back through potential node sizes to find a valid footer,
+    // The only time this would fail would be for very old indexes which have no trailing footer.
+    if (optimisticTailRead && (actualSize > maxReadSize))
     {
-        _WINREV(hdr.nodeSize);
-        offset_t actualSize = io->size();
-        offset_t readOffset = actualSize - hdr.nodeSize;
-        sizeRead = io->read(readOffset, sizeof(hdr), &hdr);
-        if (sizeRead != sizeof(hdr))
-            throw MakeStringException(4, "Invalid key %s: failed to read trailing key header at offset %llu, read %u", name.str(), readOffset, (unsigned)sizeRead);
+        offset_t endOffset = actualSize - maxReadSize;
+        const byte * endBuffer = readCache.queryFillBuffer(endOffset, maxReadSize, io, maxReadSize);
+        assertex(endBuffer);
+
+        const byte * end = endBuffer + maxReadSize;
+        for (unsigned guessedNodeSize = 512; guessedNodeSize < 0x10000; guessedNodeSize *= 2)
+        {
+            const KeyHdr * candidateHdr = (const KeyHdr *)(end - guessedNodeSize);
+
+            //Does this look like a valid header?  Check the node size is consistent
+            unsigned short hdrNodeSize = candidateHdr->nodeSize;
+            _WINREV(hdrNodeSize);
+            if (hdrNodeSize != guessedNodeSize)
+                continue;
+
+            //Check the file size field matches the actual size of the file.
+            __int64 phyrec = candidateHdr->phyrec;
+            _WINREV(phyrec);
+            if (phyrec != actualSize-1)
+                continue;
+
+            // This should never be false - possibly an internal error instead
+            if (!(candidateHdr->ktype & USE_TRAILING_HEADER))
+                continue;
+
+            // Likelihood of a false positive by this point is not worth considering
+            memcpy(&hdr, candidateHdr, sizeof(KeyHdr));
+            foundTrailingHeader = true;
+            break;
+        }
+    }
+
+    if (!foundTrailingHeader)
+    {
+        // Read the first section of the file to prime the page cache - small indexes will be fully read
+        size32_t readSize = (size32_t)std::min(actualSize, (offset_t)maxReadSize);
+        const byte * match = readCache.queryFillBuffer(0, sizeof(KeyHdr), io, readSize);
```

Comment:
An index couldn't ever be smaller than sizeof(KeyHdr) could it ?

Reply:
no, and this code will correctly throw an error if it ever is smaller.